core/drivers/crypto/se050/core/ctr.c

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2020 - All Rights Reserved
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <assert.h>
+#include <crypto/crypto_impl.h>
+#include <se050.h>
+#include <se050_cipher_algorithms.h>
+#include <string.h>
+#include <utee_defines.h>
+#include <util.h>
+
+struct se050_aes_ctr_ctx {
+	struct crypto_cipher_ctx ctx;
+	sss_se05x_symmetric_t aes_ctx;
+	sss_se05x_object_t key_obj;
+	uint8_t *cnt;
```

Comment:
but that would be wasting memory since do_copy_state is not always/necessarily called (I'd say it is a rare call mostly used in xtest from what I have seen). Also by wasting memory we could be causing an issue somewhere else...I'd rather keep it the way it is.

Reply:
why would do_copy_state segfault?...nothing would happen (huh, not even an EMSG...need to add that)

also remember that the counter 'object' is shared between all copied contexts (it doesnt belong to just one context); notice that on destruction of a context, the counter needs to be updated on all the other copied contexts to know which one will have to do the final housekeeping....so a pointer is really the only way.

**btw the condition is actually wrong (another regression):** 
the could should be



```
static void do_final(struct crypto_cipher_ctx *ctx)
{
	struct se050_aes_ctr_ctx *c = to_aes_ctr_ctx(ctx);

	if (!se050_refcount_final_ctx(c->cnt)) {
		memset(c->block, 0, sizeof(c->block));
		return;
	}

	if (c->key_obj.keyId)
		sss_se05x_key_store_erase_key(se050_kstore, &c->key_obj);

	sss_se05x_symmetric_context_free(&c->aes_ctx);
}
```

Reply:
Oh, I missed that. I should take more time to read that code.
Here I really think you need some inline comment to explicitly state this is a shared counter for context in case it's duplicated.

> why would do_copy_state segfault?...nothing would happen (huh, not even an EMSG...need to add that)

No segfault? Assume `do_copy_state()` fails to allocate the counter. When one of the context is freed, `sss_se05x_symmetric_context_free(&c->aes_ctx);` is called. What would happen then to the other context when it tries to use the freed `c->aes_ctx`?  Hmmm, maybe nothing but reporting a error. If so, 
then fine.